agent/agent_test.go

```diff
@@ -5328,9 +5328,421 @@ func uniqueAddrs(srvs []apiServer) map[string]struct{} {
 	return result
 }
 
-func runStep(t *testing.T, name string, fn func(t *testing.T)) {
-	t.Helper()
-	if !t.Run(name, fn) {
-		t.FailNow()
+func TestAgent_AutoReloadConfigEnable(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	t.Parallel()
+
+	a := NewTestAgent(t, `auto_reload_config = true`)
+	defer a.Shutdown()
+
+	require.True(t, a.Agent.config.AutoReloadConfig)
+}
+
+func TestAgent_AutoReloadConfigDefaultDisable(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	t.Parallel()
+
+	a := NewTestAgent(t, ``)
+	defer a.Shutdown()
+
+	require.False(t, a.Agent.config.AutoReloadConfig)
+}
+
+func TestAgent_AutoReloadDoReload_WhenCertAndKeyUpdated(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
```

Comment:
I know that a _lot_ of tests in this package get skipped locally because their side-effects (e.g. writing to the filesystem) make them slow. But I wonder if there's an opportunity to decrease some coupling and improve testability here, for example by making the `FileWatcher` an interface and using an in-memory implementation of [`io/fs`](https://pkg.go.dev/io/fs).

Maybe a bit out of the scope of this PR, so don't feel like you need to address it. I just feel a bit uncomfortable about how deeply we rely on CI and actual sleeping.

Reply:
I converted the Watcher into an interface for the purpose of implementing a rate limited watcher for the time being. I will revisit those tests later to remove some of them in favour of tests with a mock watcher (I will probably keep 1 or 2 to make sure the whole thing work though)